src/bzflag/ServerLink.cxx

```diff
@@ -491,8 +482,69 @@ void            ServerLink::send(uint16_t code, uint16_t len,
 #endif
 #endif //WIN32
 
+int ServerLink::fillTcpReadBuffer()
+{
+    int emptySpace = sizeof(tbuf) - tcpBufferPos;
+    if (!emptySpace)
+        return 0;
+    ssize_t rlen = recv(fd, &tbuf[tcpBufferPos], emptySpace, 0);
+    if (rlen < 0)
+    {
+        if (errno == EAGAIN)
+            return 0;
+        else if (errno == EWOULDBLOCK)
+            return 0;
+        else
+            return -1;
+    }
+
+    if (rlen)
+    {
+#if defined(NETWORK_STATS)
+        bytesReceived += rlen;
+        packetsReceived++;
+#endif
+        tcpBufferPos += rlen;
+    }
+    return 0;
+}
+
+void ServerLink::consumeTcpReadBuffer(unsigned int len)
+{
+    int remain  = tcpBufferPos - len;
+
+    if (remain)
+        memmove(tbuf, &tbuf[len], remain);
+    tcpBufferPos -= len;
+}
+
+// A packet is there?
+bool ServerLink::tcpPacketIn()
+{
+    // Enough data to decode the header?
+    if (tcpBufferPos < 4)
+        return false;
+
+    // unpack len
+    const void* buf = tbuf;
+    uint16_t    tmpLen;
+    buf = nboUnpackUShort(buf, tmpLen);
+
+    // Add header size
+    tmpLen += 4;
+
+    // Check if len is good
+    if (tmpLen > MaxPacketLen)
+        // Len is bigger than the Max Packet Len
+        // The packet is there but it is wrong
+        return true;
```

Comment:
again, don't be shy on braces with multi-line conditions. 

If a client sends a packet that is malformed (packet len too long is malicious), then we should reject the connection. Our protocol expects people to follow it. It's appropriate to validate all input, but "bad" input should be aggressively rejected.

Reply:
that is hardly rejected in the caller. Maybe an exception could be more appropriate, but we never use exception.

The line that handle it is 

    // unpack len and message code
    const void* buf = tbuf;
    buf = nboUnpackUShort(buf, len);
    buf = nboUnpackUShort(buf, code);

    **if (len > MaxPacketLen + 4)
    {
        // Len is bigger than the Max Packet Len
        return -1;
    }**